Server/mods/deathmatch/logic/CObjectSync.cpp

```diff
@@ -196,43 +193,106 @@ void CObjectSync::Packet_ObjectSync(CObjectSyncPacket& Packet)
+        if (pData->ucFlags & 0x10)
+        {
+            float fPreviousHealth = pObject->GetHealth();
+            pObject->SetHealth(pData->fHealth);
 
-            // Grab the element
-            CElement* pElement = CElementIDs::GetElement(pData->ID);
-            if (pElement && IS_OBJECT(pElement))
+            if (pData->fHealth < fPreviousHealth)
             {
-                CObject* pObject = static_cast<CObject*>(pElement);
+                CElement* pAttacker = CElementIDs::GetElement(pData->attackerID);
 
-                // Is the player syncing this object?
-                if ((pObject->GetSyncer() == pPlayer) && pObject->CanUpdateSync(pData->ucSyncTimeContext))
+                if (pData->fHealth > 0)
                 {
-                    // Apply the data to the object
-                    if (pData->ucFlags & 0x1)
+                    float fLoss = fPreviousHealth - pData->fHealth;
+
+                    if (fLoss > 0)
                     {
-                        pObject->SetPosition(pData->vecPosition);
-                        g_pGame->GetColManager()->DoHitDetection(pObject->GetPosition(), pObject);
+                        CLuaArguments Arguments;
+                        Arguments.PushNumber(fLoss);
+
+                        if (pAttacker)
+                            Arguments.PushElement(pAttacker);
+                        else
+                            Arguments.PushNil();
+
+                        pObject->CallEvent("onObjectDamage", Arguments);
                     }
-                    if (pData->ucFlags & 0x2)
-                        pObject->SetRotation(pData->vecRotation);
-                    if (pData->ucFlags & 0x4)
-                        pObject->SetHealth(pData->fHealth);
+                }
+                else // object has been break?
+                {
+                    CLuaArguments Arguments;
+                    if (pAttacker)
+                        Arguments.PushElement(pAttacker);
+                    else
+                        Arguments.PushNil();
 
-                    // Send this sync
-                    pData->bSend = true;
+                    pObject->CallEvent("onObjectBreak", Arguments);
                 }
             }
         }
+        if (pData->ucFlags & 0x20)            // Sync inWater state from client
+            pObject->SetInWater(pData->bIsInWater);
+        // Sync properties
+        if (pData->ucFlags & 0x40)
+            pObject->SetMass(pData->fMass);
+        if (pData->ucFlags & 0x80)
+            pObject->SetTurnMass(pData->fTurnMass);
+        if (pData->ucFlags & 0x100)
+            pObject->SetAirResistance(pData->fAirResistance);
+        if (pData->ucFlags & 0x200)
+            pObject->SetElasticity(pData->fElasticity);
+        if (pData->ucFlags & 0x400)
+            pObject->SetBuoyancyConstant(pData->fBuoyancyConstant);
+        if (pData->ucFlags & 0x800)
+            pObject->SetCenterOfMass(pData->vecCenterOfMass);
```

Comment:
I think we don't need to sync static data

Reply:
Data is only sent if it has changed. If the object is created on the server side, you must first download its properties, which are available on the client side, otherwise the properties on the server and client side would be different

Reply:
But why do we need that? 
If you create object on server, you should change properties on server.

Reply:
This is illogical. When creating an object on the server side, its properties will be default (-1, zero vector), and on the client side the properties will have some values.

When we create a vehicle on the server side, it also has handling, plate and other data just like on the client side. It's as if upgrades were only on the client side by default, and there were no upgrades on the server side because they had to be set on the server side as well.

This is strange behavior imo

Reply:
> When creating an object on the server side, its properties will be default (-1, zero vector)

So server should have a config. All object properties should be available for a new object immediately.

Reply:
If you know where and how to do it in a different way as you write, give me a hint and tell me

Reply:
GTA has 160 physical groups. You can store them on server and add a hashmap `modelId` -> `groupId`
You can use `engineGetObjectGroupPhysicalProperty` and `engineGetModelPhysicalPropertiesGroup` to generate data.
Horewer, I have no idea what we should return for objects without a physical group. Does gta use some default parameters for them?

Reply:
This requires a lot of additional work for the same effect, which is not profitable. Why waste time doing something in a more complex way when the current way works well and doesn't cause problems? I see no reason to waste time doing the same thing in a more difficult way

Reply:
No, your solution adds uninitialized state for server objects. So it's much complex, that the solution with serverside config

Reply:
This is normal behavior. The object has its data set at the time of synchronization and this is normal, because on the client side if the object is not streamed, its properties are also default (-1, zero vector). The properties have the appropriate values ​​when the object is streamed, and if the object is streamed, there is a good chance that it is also synchronized. I don't understand why you should complicate such a trivial thing as object properties. If this is really such a bad way, then I will remove access to the server-side properties, because implementing the same in a more difficult way requires investigation and a lot of work for the same effect

Reply:
This is not normal and unnecessary complex

Reply:
> This is not normal and unnecessary complex

Adding feedback label due to this, cc @FileEX to please provide counterarguments or chat with eachother in dev discord about your POV's. Only after that comes to a conclusion, this can be marked as resolved.

Reply:
I will work on it. Now I'm breaking the entire PR into several smaller ones to make it easier to review. One of them is already ready and waiting to be merged #3450